Cleanup itemframes

Postremus opened the pull request:
Moved facing attribute into new GlowHangingEntity super class.
Remove all methods used to set or get by a "facingNumber", and use the new enum HangingFace instead.
Pull save and load of "Facing" attribute up into new super class "HangingStore".
call some more of the hanging* events, but there are for sure some instances missing where more events should be called.

related to #504, since this is the groundwork for paintings

Postremus committed:
Cleanup itemframes

Moved facing attribute into new GlowHangingEntity super class.
Remove all methods to set or get by a "facingNumber", and use the new enum HangingFace instead
Pull "HangingEntity" save and load up into new super class "HangingStore"
call some more of the hanging* events, but there are for sure some instances missing where more events should be called

related to #504, since this is the groundwork for paintings

Postremus committed:
make byBlockFace map static final

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowItemFrame.java:
```diff
@@ -249,60 +177,46 @@ public Location getOrigin() {
     public void setItem(ItemStack is) {
-        setItemInFrame(is);
+        if (is == null) {
+            is = new ItemStack(Material.AIR, 1);
```
You can use `is = InventoryUtil.itemOrEmpty(is)`

Postremus committed:
Drop item naturally
when attacking the frame
itemframes are also removed because of physics -> call hangingbreakevent
fix possible duplication of items in frame

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowItemFrame.java:
```diff
@@ -196,10 +200,10 @@ public void setFacingDirection(BlockFace blockface) {
-        }
+        is = InventoryUtil.itemOrEmpty(is);
+        is = is.clone();
         is.setAmount(1);
```
I'm not sure what you're doing here. If a plugin uses this method with say, `ItemStack(Material.STONE, 0)` it should be treated as empty instead of updating its amount to 1 (unless vanilla does it differently?). Cloning the ItemStack should not be necessary either.

Postremus replied:
Cloning is required, imo.
During testing, I had the following scenario:
- I have 4 Rails in my Inventory
- I put one into the itemframe. Now I have 3 Rails in my Inventory.
- I rightclick onto the itemframe, and the items drop.
- I pick them up, and now I have 6 Rails in my Inventory.
Cloning solved this scenario.

Also, at least on paper, the amount of the item seems to be always set to 1.

aramperes replied:
Alright. You can still simplify this with `is = InventoryUtil.itemOrEmpty(is).clone()`.

Postremus committed:
simplify setItem

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowItemFrame.java:
```diff
@@ -44,81 +43,21 @@ public GlowItemFrame(GlowPlayer player, Location location, BlockFace clickedface
-                return BlockFace.SOUTH;
-        }
+        setRotation(Rotation.NONE);
+        setItem(new ItemStack(Material.AIR));
```
`InventoryUtil.createEmptyStack()`

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowItemFrame.java:
```diff
@@ -44,81 +43,21 @@ public GlowItemFrame(GlowPlayer player, Location location, BlockFace clickedface
         if (message.getAction() == Action.INTERACT.ordinal()) {
-            if (itemInFrame == Material.AIR) {
+            if (InventoryUtil.isEmpty(getItem())) {
                 ItemStack isInHand = player.getItemInHand();
```
Please take the hand into consideration -- see https://github.com/GlowstoneMC/Glowstone/blob/master/src/main/java/net/glowstone/net/message/play/player/InteractEntityMessage.java#L25 and https://github.com/GlowstoneMC/Glowstone/blob/master/src/main/java/net/glowstone/net/handler/play/player/InteractEntityHandler.java#L52

aramperes replied:
I'll correct myself: it doesn't seem like you can place items from the off-hand for Item Frames and Armor Stands (see: https://bugs.mojang.com/browse/MC-84771). Please add a comment or `todo` around this check in case this vanilla bug is fixed in the future (although it has been opened for nearly 2 years). The check remains OK as long as they don't fix this.

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowItemFrame.java:
```diff
@@ -129,19 +68,21 @@ public boolean entityInteract(GlowPlayer player, InteractEntityMessage message)
-                setItemInFrame(new ItemStack(Material.AIR));
-                rot = 0;
+                world.dropItemNaturally(location, getItem().clone());
+                setItem(new ItemStack(Material.AIR));
```
`InventoryUtil.createEmptyStack()`

Postremus committed:
use InventoryUtil to create empty item stacks

Momothereal committed:
Merge branch 'master' into #504-itemframe
